### PR TITLE
feat: port the node task to hardhat v3

### PR DESCRIPTION
**Pull request by galargh**
- [ ] Because this PR includes a **bug fix**, relevant tests have been included.
- [ ] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.
- [x] I didn't do anything of this.

---

In this PR, I port the node task from v2 to v3. I describe my design choices in https://www.notion.so/nomicfoundation/Node-Task-154578cdeaf5802c870ee4c7ea7a05ba (this doc should be helpful in guiding the review)

## Follow-ups

- https://github.com/NomicFoundation/hardhat/issues/6040
- https://github.com/NomicFoundation/hardhat/issues/6042

## Testing

1. I successfully started a node using `hardhat3 node` and deployed a contract from the v2 sample project to it using ignition.
2. I added a test case in which I start a JSON RPC server from an EDR provider, create an additional HTTP provider pointing at the server, and compare responses to an `eth_chainId` request. The test suite could definitely be expanded.

### CODE REVIEW

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts
Comment (galargh) [inline v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts:1]: This is a direct port of https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/internal/hardhat-network/jsonrpc/handler.ts

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/server.ts
Comment (galargh) [inline v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/server.ts:1]: This is a direct port of https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/internal/hardhat-network/jsonrpc/server.ts

**Comment by galargh** [comment]
@schaable Tagging you for a review, too, as per our initiative to spread the process. https://www.notion.so/nomicfoundation/Node-Task-154578cdeaf5802c870ee4c7ea7a05ba should be helpful for a little more context.

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+} from "../../network-manager/edr/errors.js";
+import {
+  isJsonRpcRequest,
+  isJsonRpcResponse,
+  isSuccessfulJsonRpcResponse,
+} from "../../network-manager/json-rpc.js";
+import { ProviderError } from "../../network-manager/provider-errors.js";
+
+export class JsonRpcHandler {
+  readonly #provider: EIP1193Provider;
```
Comment (schaable) [inline v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts:25]: I believe this should be an `EthereumProvider`

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+const _handleError = (error: any): JsonRpcResponse => {
+  // extract the relevant fields from the error before wrapping it
+  let txHash: string | undefined;
+  let returnData: string | undefined;
+
+  if (error.transactionHash !== undefined) {
+    txHash = error.transactionHash;
+  }
+  if (error.data !== undefined) {
```
Comment (schaable) [inline v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts:254]: minor: the optional chaining operator shouldn't be necessary within this code block.

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+      code: error.code,
+      message: error.message,
+    },
+  };
+
+  const data: any = {
+    message: error.message,
+  };
+
+  if (txHash !== undefined) {
```
Comment (schaable) [inline v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts:284]: why is this block empty?

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+  #sendResponse(
+    res: ServerResponse,
+    rpcResp: JsonRpcResponse | JsonRpcResponse[],
+  ) {
+    res.statusCode = 200;
+    res.setHeader("Content-Type", "application/json");
+    res.end(JSON.stringify(rpcResp));
+  }
+
+  async #handleSingleRequest(req: JsonRpcRequest): Promise<JsonRpcResponse> {
```
Comment (schaable) [inline v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts:149]: minor: I’d suggest moving the `handleRequest` function code into this function and renaming this one to `handleRequest`. Having both can be a bit confusing.

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+      rpcResp = _handleError(error);
+    }
+
+    // Validate the RPC response.
+    if (!isJsonRpcResponse(rpcResp)) {
+      // Malformed response coming from the provider, report to user as an internal error.
+      rpcResp = _handleError(new InternalError("Internal error"));
+    }
+
+    if (rpcReq !== undefined) {
```
Comment (schaable) [inline v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts:169]: Isn’t this guaranteed to be non-undefined due to the earlier `isJsonRpcRequest` check?

**Review by schaable — COMMENTED** [review]
Hey Piotr! Overall, this looks good, but I’ve left a couple of comments. Feel free to make a decision on the ones marked as minor and consider them resolved.

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+      result,
+    };
+  };
+}
+
+const _readJsonHttpRequest = async (req: IncomingMessage): Promise<any> => {
+  let json;
+
+  try {
+    const buf = await getRawBody(req);
```
[the remark this answers is not in the document]
Comment (galargh) [reply v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts:215]: Sounds great, I'll make the change and add something like this to the utils!

[the remark this answers is not in the document]
Comment (galargh) [reply v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts:215]: It seems we can do it even simpler:
```
    const bytes: number[] = [];
    for await (const chunk of req) {
        bytes.push(...chunk);
    }
    const text = new TextDecoder("utf-8").decode(new Uint8Array(bytes));
```

I was thinking of putting this inline for now and abstracting it later if/when we have more usecases for this. Do you think that's a good idea?

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/server.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+export interface JsonRpcServerConfig {
+  hostname: string;
+  port: number;
+
+  provider: EIP1193Provider;
+}
+
+export class JsonRpcServer implements IJsonRpcServer {
```
[the remark this answers is not in the document]
Comment (galargh) [reply v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/server.ts:28]: Good catch! I copied it from v2, but it totally makes sense. I'll add it to our eng guidelines, too.

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts (hunk 3 above)
Comment (galargh) [reply v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts:284]: A great catch, I have no idea how that happened.

**Comment by galargh** [comment]
Thanks @schaable - that was really helpful 🙇

File: v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (schaable) [reply v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts]: I think it's a good idea. Go for it!

**Review by schaable — APPROVED** [review]
LGTM!

**Pull request merged**